File: dsa/hw01/code/src/uniqueint.py

import os

class UniqueInt:

    OFFSET = 1023
    RANGE = 2047 + 1

    def __init__(self):
        self.seen = [False] * UniqueInt.RANGE

    def processFile(self, inputFilePath, outputFilePath):
        self.seen = [False] * UniqueInt.RANGE

        with open(inputFilePath, 'r') as infile:
            for line in infile:
                item = self.readNextItemFromLine(line)
                if item is not None:
                    index = item + UniqueInt.OFFSET
                    self.seen[index] = True

        with open(outputFilePath, 'w') as outfile:
            for i in range(UniqueInt.RANGE):
                if self.seen[i]:
                    outfile.write(f"{i - UniqueInt.OFFSET}\n")

        print(f"Loaded: {inputFilePath}")
        print(f"Saved to: {outputFilePath}")

    def readNextItemFromLine(self, line):
        stripped = line.strip()
        if stripped == "":
            return None

        parts = stripped.split()
        if len(parts) != 1:
            return None

        try:
            num = int(parts[0])
            if num < -1023 or num > 1023:
                return None
            return num
        except ValueError:
            return None


if __name__ == "__main__":
    input_folder = "../../sample_inputs"
    output_folder = "../../sample_results"

    processor = UniqueInt()

    for file_name in os.listdir(input_folder):
        if file_name.endswith(".txt"):
            input_path = os.path.join(input_folder, file_name)
            output_file = file_name + "_results.txt"
            output_path = os.path.join(output_folder, output_file)
            processor.processFile(input_path, output_path)

File: dsa/hw01/code/src/uniqueint.js
const fs = require("fs");
const readline = require("readline");

class UniqueInt {
    constructor() {
        this.min = -1023;
        this.max = 1023;
        this.range = this.max - this.min + 1;
        this.seen = new Array(this.range).fill(false);
    }

    readNextItemFromLine(line) {
        const trimmed = line.trim();
        const parts = trimmed.split(/\s+/);

        if (parts.length !== 1) return null;
        const num = parseInt(parts[0], 10);
        if (isNaN(num) || num < this.min || num > this.max || parts[0].includes('.')) {
            return null;
        }
        return num;
    }

    processFile(inputFilePath, outputFilePath) {
        const data = fs.readFileSync(inputFilePath, "utf-8").split("\n");
        for (let line of data) {
            const num = this.readNextItemFromLine(line);
            if (num !== null) {
                this.seen[num - this.min] = true;
            }
        }

        const uniqueSorted = [];
        for (let i = 0; i < this.range; i++) {
            if (this.seen[i]) {
                uniqueSorted.push(i + this.min);
            }
        }

        fs.writeFileSync(outputFilePath, uniqueSorted.join("\n"));

        console.log(`Processed: ${inputFilePath}`);
        console.log(`Output: ${outputFilePath}`);
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question("Enter input file path: ", (inputFile) => {
    const outputFile = "output.txt"; 
    const uniqueInt = new UniqueInt();
    try {
        uniqueInt.processFile(inputFile, outputFile);
    } catch (error) {
        console.error("Error:", error.message);
    }
    rl.close();
});
